---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```
#CIDR - Clustering through Imputation and Dimensionality Reduction

Fast and accurate clustering through imputation and dimensionality
reduction for single cell RNA-seq data.

Most existing dimensionality reduction and clustering packages for single cell RNA-Seq (scRNA-Seq) data deal with dropouts by heavy modelling and computational machinery. Here we introduce _CIDR_ (Clustering through Imputation and Dimensionality Reduction), an ultrafast
algorithm which uses a novel yet very simple ‘implicit imputation’ approach to alleviate the
impact of dropouts in scRNA-Seq data in a principled manner.

For more details about _CIDR_ refer to the [preprint](http://biorxiv.org/content/early/2016/08/10/068775) on BioRxiv.

##Getting Started

* Make sure your version of R is at least 3.1.0.
* _CIDR_ has been tested primarily on the Linux and Mac platforms.  _CIDR_ has also been tested on the Windows platform - however this requires the use of an external software package _Rtools_.
* If you are on the Windows platorm, ensure that [Rtools](https://cran.r-project.org/bin/windows/Rtools/) is installed. Rtools is software (installed external to R) that assists in building R packages, and R itself.  Note that the downlaod for _Rtools_ is in the order of 100M.
* Install the CRAN package _devtools_ package which will be used to install _CIDR_ and its dependencies:
```{r, eval=F}
## this is an R command
install.packages("devtools")
```

* Install the _CIDR_ package directly from the Github repository (including any dependencies):

```{r, eval=F}
## this is an R command
devtools::install_github("VCCRI/CIDR")
## Note that for some Windows platforms, you may be asked to re-install RTools
## - even though it may already have been installed.  Say yes if prompted.
## Your windows platform may require the specific version of RTools being suggested.
```

#Examples
##Simulated Data
Test the newly installed _CIDR_ package:

```{r}
library(cidr)
example("cidr")
```

##Biological Datasets
Further examples of applying _CIDR_ using real biological datasets can be found at this [Github repository](https://github.com/VCCRI/CIDR-examples).  The name of the repository is _CIDR-examples_. 

Clicking on the _Clone or Download_ button in the Github repository for _CIDR-examples_ will enable the user to download a zip file containing the raw biological data and the R files for the examples.  The user can then extract the files and run the provided R examples.

###Human Brain scRNA-Seq Dataset
_CIDR-examples_ contains a human brain single-cell RNA-Seq dataset, located in the _Brain_ folder. In this dataset
there are 420 cells in 8 cell types after we exclude hybrid cells.  

Reference for human brain dataset:

Darmanis, S. _et al._ A survey of human brain transcriptome diversity at the single cell level.
_Proceedings of the National Academy of Sciences_ 112, 7285–7290 (2015).

###Human Pancreatic Islet scRNA-Seq Dataset
_CIDR-examples_ contains a human pancreatic islet single-cell RNA-Seq dataset, located in the _PancreaticIslet_ folder. In this dataset there are 60 cells in 6 cell types after we exclude undefined cells and bulk RNA-Seq samples.  

Reference for human pancreatic islet dataset:

Li, J. _et al._ Single-cell transcriptomes reveal characteristic features of human pancreatic islet
cell types. _EMBO Reports_ 17, 178–187 (2016).

